### ISSUE TITLE
'ended' event not firing

### ISSUE THREAD

**Issue by sanbornhnewyyz** [google/shaka-player#220]
This is not strictly speaking a Shaka issue as far as I can tell but I was hoping you might be able to provide insight. 

We are having an issue where the 'ended' event on the <video> element is not firing when our content finishes. Actually it seems some of our content is triggering an 'ended' event sometimes and sometimes it's not but I can't figure out if there is a pattern to it. Using the Shaka sample app I loaded some content and played it until the end and no 'ended' events fired. After reloading the page and loading the same content the 'ended' events were firing.

I can see that Shaka has it's own 'streamEnded' event that fires close to the end of the content to trigger mediaSource.endOfStream(). Is there any relation between this event and the <video> tag 'ended' event? The streamEnded event is always firing so clearly Shaka is consistently aware that the end of the content has been reached but for some reason which I'm having a really hard time tracking down, the <video> tag ended condition isn't getting met. I suspect this is a content issue but that's as far as I am on this right now.

**Comment by ddorwin** [comment]
If this is in Chrome, it might be related to https://code.google.com/p/chromium/issues/detail?id=349543.

**Comment by joeyparrish** [comment]
TL;DR: @ddorwin is probably right.

This is how end-of-stream works in Shaka:

When a Stream appends the final segment for its content type (audio or video), it fires an internal, non-bubbling 'streamEnded' event.  This goes to the StreamVideoSource, which owns the MediaSource object.

When all Stream objects have reported 'streamEnded', then final segments have been appended for all types, and StreamVideoSource calls MediaSource.endOfStream().  This tells the browser's video stack that we will not be providing any additional segments.

When MediaSource.endOfStream() has been called AND video.currentTime hits the end of the timeline, then the video element is supposed to fire the 'ended' event.  (If we didn't call endOfStream(), the video would simply pause at the end, waiting for more data to extend the video timeline.)

So if 'ended' isn't fired on the video, it's very likely a browser bug, and it seems @ddorwin knows a likely candidate.

You could patch Shaka to work around the Chrome bug by emitting its own event when MediaSource.endOfStream() and Math.abs(video.currentTime - video.duration) < someTinyThreshold.  This, however, would not be something we would accept as a patch.  We try not to hide browser bugs if we can help it, in particular when the impact is this small.

**Comment by sanbornhnewyyz** [comment]
Great, thanks for all the info guys, this is really helpful. I hadn't seen that chrome bug. We are indeed experiencing this in chrome.

It had occurred to me to listen to this endOfStream event but I agree it's a bad solution and if we absolutely need a temp solution then we will just do it ourselves.

**Comment by ddorwin** [comment]
Which version of Chrome? One of the most recent comments says "Remaining cases of this seem to be WebAudio issues...", which would imply Shaka should not be impacted. If you can provide a repro case or more details in that bug, that would be helpful.

**Comment by sanbornhnewyyz** [comment]
I'm running chrome: 46.0.2490.80 (64-bit) and chrome canary:  48.0.2550.0 canary (64-bit) also seems to behave the same for me. This is on MAC OSX 10.9.5

There is audio associated with the content though so couldn't it still be related to these unresolved webAudio issues?

In any case, I'll see if I can get some content that exhibits this issue that we can make public and then I can put reproduction steps on that issue.

**Comment by joeyparrish** [comment]
Thanks @ddorwin for pointing us at the Chrome bug, and thanks @sanbornhnewyyz for offering to contribute to that report.  Since it appears this isn't a Shaka bug, I'm going to go ahead and close this issue.
